Workaround to avoid crashing bbPress' frontend profile page

CryoutCreations opened the pull request:
This workarounds skips the addition of the TwoFactor fields to the user profile when it's part of bbPress' frontend profile edit page. These are settings that shouldn't probably by editable on the forum profile anyway. 

Fixes the `Fatal error: Uncaught Error: Call to undefined function convert_to_screen()` reported in #175

kasparsd commented on class-two-factor-core.php:
```diff
@@ -925,6 +925,11 @@ public static function manage_users_custom_column( $output, $column_name, $user_
 	public static function user_two_factor_options( $user ) {
+		
+		/* workaround for two factor options crashing bbpress' profile page */
+		if ( function_exists('bbp_is_user_home_edit') && bbp_is_user_home_edit() ) return;
```
This is currently relying on a specific integration and historically this plugin doesn't have any conditional logic to account for specific integrations. Could we somehow convert this into something that can be placed in the `class-two-factor-compat.php` file which currently deals with Jetpack specific things?

https://github.com/WordPress/two-factor/blob/master/class-two-factor-compat.php

Which of the functionality in this method is actually not compatible with the front-end output?

r-a-y replied:
You might want to check out my comment where I outline frontend integration with this problem (and a few more) here: https://github.com/WordPress/two-factor/issues/335#issuecomment-719087800

For this specific issue, it would probably be better to redeclare the admin functions on the frontend when needed.

CryoutCreations replied:
> 
> 
> This is currently relying on a specific integration and historically this plugin doesn't have any conditional logic to account for specific integrations. Could we somehow convert this into something that can be placed in the `class-two-factor-compat.php` file which currently deals with Jetpack specific things?
> 
> https://github.com/WordPress/two-factor/blob/master/class-two-factor-compat.php
> 
> Which of the functionality in this method is actually not compatible with the front-end output?

I updated the PR to include an example error screenshot.
Until the required admin functions are exposed on the frontend (should they?), it might be possible to rewrite the workaround in the compat file to conditionally remove the two action hooks instead. I'll look into this.

CryoutCreations commented:
Per @kasparsd's request I moved the bbPress conditionality check to the `two-factor-compat` class and changed the function call to use a more general function. 

Hopefully the code now meets the logic separation requirements.

jeffpaul commented:
Per [yesterday's bug scrub](https://wordpress.slack.com/archives/C04L4KCUB/p1648040434425739), while we want to be compatible with official plugins we're going to punt this to a future release so that the 0.8.0 release can focus on the U2F deprecation.

jeffpaul commented:
@georgestephanis perhaps we close this PR and https://github.com/WordPress/two-factor/issues/175 and allow things to get resolved within bbp?

masteradhoc commented:
@georgestephanis @jeffpaul isnt the reason here that `WP_List_Table` is called? We dont call `convert_to_screen` directly but over `WP_List_Table` its called.

If i see `WP_List_Table` is called only on our fido implementation, [see here:](https://github.com/search?q=repo%3AWordPress%2Ftwo-factor%20WP_List_Table&type=code)

Given that this is removed soon (#439) i think its good to close this and focus on #335 to build a proper integration option.
